subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalTasksIntegrationTest.groovy

```diff
@@ -108,4 +113,124 @@ class IncrementalTasksIntegrationTest extends AbstractIntegrationSpec {
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/9320")
+    def "incremental task with NAME_ONLY input (matching file names and content) detects moved files (type: #taskChangeType.simpleName)"() {
+        def taskName = "incrementalTask"
+        def inputs = folderNames().collect { file("${it}/input.txt").createFile() }
+        def movableInput = inputs[1]
+        def renamedInput = file("moved/${movableInput.name}")
+
+        @Language("Groovy") script = """
+            class IncrementalTask extends DefaultTask {
+                @InputFiles
+                @PathSensitive(PathSensitivity.NAME_ONLY)
+                def inputFiles = project.files(${asFileList(inputs)}, '${renamedInput}')
+            
+                @TaskAction
+                def action(${taskChangeType.name} changes) {}
+            }
+            
+            task ${taskName}(type: IncrementalTask) {}
+        """
+        buildScript script
+        run taskName, "--info"
+
+        when:
+        println "Moving ${movableInput.absolutePath} to '${renamedInput.absolutePath}'!"
+        renamedInput.text = movableInput.text
+        movableInput.delete()
+
+        run taskName, "--info"
+        then:
+        !movableInput.exists()
+        renamedInput.exists()
+        outputContains "is up-to-date"
+
+        where:
+        taskChangeType << [IncrementalTaskInputs, InputChanges]
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/9320")
+    def "incremental task with NAME_ONLY inputs (matching file names and content) detects deleted file (type: #taskChangeType.simpleName)"() {
+        def taskName = "incrementalTask"
+        def inputs = folderNames().collect { file("${it}/input.txt").createFile() }
+
+        @Language("Groovy") script = """
+            class IncrementalTask extends DefaultTask {
+                @InputFiles
+                @PathSensitive(PathSensitivity.NAME_ONLY)
+                def inputFiles = project.files(${asFileList(inputs)})
+            
+                @TaskAction
+                def action(${taskChangeType.name} changes) {}
+            }
+            
+            task ${taskName}(type: IncrementalTask) {}
+        """
+        buildScript script
+        run taskName
+
+        when:
+        inputs[1..2]*.delete()
+        println "${inputs} exists: ${inputs*.exists()}"
+        run taskName, "--info"
+        then:
+        outputDoesNotContain "${inputs[0]} has been removed."
+        outputContains "${inputs[1]} has been removed."
+        outputContains "${inputs[2]} has been removed."
+
+        where:
+        taskChangeType << [IncrementalTaskInputs, InputChanges]
+    }
+
+    private static Range<String> folderNames() { 'a'..'c' }
+
+    private static String asFileList(List<TestFile> inputs) {
+        inputs.collect { "'${normalizePath(it)}'" }.join(", ")
+    }
+
+    private static String normalizePath(TestFile it) {
+        it.absolutePath.replace('\\', '\\\\')
```

Comment:
Note that there is already `TextUtil.escapeString()`, which does what you want.

Reply:
Thanks